Refactor rustfmt's interface and test code

marcusklaas opened the pull request:
In this PR, we propose a new public function `format`, which simply returns the result of the reformatting, without any side effects. This should give us more freedom and control than wrapping logic in a callback.

We use this extra control to update the test code, which no longer needs to rely on `thread::catch_panic`, and run some extra checks. From now on, when rustfmt's own code exceeds the maximum_width, the tests will no longer pass.

Fixes https://github.com/nrc/rustfmt/issues/183.

nrc commented on src/comment.rs:
```diff
@@ -288,7 +288,9 @@ impl<T> Iterator for CharClasses<T> where T: Iterator, T::Item: RichChar {
 mod test {
     use super::{CharClasses, CodeCharKind, contains_comment, rewrite_comment, FindUncommented};
 
+    // TODO(#217): prevent string literal from going over the limit.
```
Use FIXME instead of TODO (we should work to make rustfmt TODO free soon)

nrc commented on src/comment.rs:
```diff
@@ -300,9 +302,10 @@ mod test {
+        let expected =
+            "/* com\n                                                                      \
+             * men\n                                                                      \
+             * t */";
```
I think you must change this statement in every PR you send :-)

nrc commented:
sweet! r+ with the TODO nit (in general I think any TODO with an issue number should be a FIXME, but you don't need to change all the others right now).

marcusklaas commented:
Updated.

I introduced a small unsafe block to prevent cloning of the config structs. Could someone have a quick look at this to make sure this is OK?

https://github.com/nrc/rustfmt/pull/326/files#diff-b4aea3e418ccdb71239b96952d9cddb6R316

nrc commented on src/lib.rs:
```diff
@@ -310,8 +274,8 @@ fn fmt_lines(file_map: &mut FileMap, config: &Config) -> FormatReport {
 struct RustFmtCalls {
-    write_mode: WriteMode,
-    config: Option<Box<config::Config>>,
+    config: &'static Config,
```
could you use a lifetime variable rather than 'static here?

nrc commented:
I think the unsafe block is ok, but it would be clearer to parameterise the RistFmtCalls object by a lifetime and transmute that, rather than the config.

nrc commented:
TBH, I don't think a single clone of the config struct is a big deal, and probably better than having an unsafe block. I'd only worry if it were in a tight loop or something.

marcusklaas commented:
I've tried that, but it doesn't work since `build_controller` takes a `&mut self`. The compiler won't let us take an explicit lifetime there.

marcusklaas commented:
Woops, I misunderstood your comment. I like your first option more than the clone. Since we'd also need `Rc`s in that case.

marcusklaas commented:
After some reconsideration, I now share the view that being `unsafe` free is probably worth more than saving a single clone and deref.
